### PR TITLE
feat: introduce identity topup transition

### PR DESCRIPTION
**Pull request by antouhou**

### CODE REVIEW

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const ValidationResult = require('../../../validation/ValidationResult');
+
+const IdentityNotFoundError = require('../../../errors/IdentityNotFoundError');
+
+/**
+ * @param {StateRepository} stateRepository
+ * @param {validateLockTransaction} validateLockTransaction
+ * @return {validateIdentityTopUpSTData}
+ */
+function validateIdentityTopUpSTDataFactory(stateRepository, validateLockTransaction) {
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js:10]: validateIdentityTopUpTransitionDataFactory

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * @typedef validateIdentityTopUpSTData
+   * @param {IdentityTopUpTransition} identityTopUpTransition
+   * @return {ValidationResult}
+   */
+  async function validateIdentityTopUpSTData(identityTopUpTransition) {
+    const result = new ValidationResult();
+
+    // Check if identity with such id already exists
+    const identityId = identityTopUpTransition.getIdentityId();
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js:27]: Please use `validateIdentityExistence`

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+const ValidationResult = require('../../../validation/ValidationResult');
+const IdentityStateTransition = require('./IdentityTopUpTransition');
+
+/**
+ * @return {validateIdentityTopUpST}
+ */
+function validateIdentityTopUpSTStructureFactory() {
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js:7]: validateIdentityTopUpTransitionStructureFactory

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    if (identityStateTransition instanceof IdentityStateTransition) {
+      rawIdentityStateTransition = identityStateTransition.toJSON();
+    } else {
+      rawIdentityStateTransition = identityStateTransition;
+    }
+
+    const result = new ValidationResult();
+
+    return result;
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js:24]: Let's just return `new ValidationResult()` ?

File: lib/test/fixtures/getIdentityTopUpSTFixture.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const IdentityTopUpTransition = require('../../identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition');
+
+const stateTransitionTypes = require('../../stateTransition/stateTransitionTypes');
+
+/**
+ *
+ * @return {IdentityTopUpTransition}
+ */
+module.exports = function getIdentityCreateSTFixture() {
```
Comment (shumkov) [inline lib/test/fixtures/getIdentityTopUpSTFixture.js:9]: getIdentityTopUpTransitionFixture

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js (hunk 1 above)
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js (hunk 3 above)
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js (hunk 4 above)
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTStructureFactory.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * @typedef validateIdentityTopUpTransitionData
+   * @param {IdentityTopUpTransition} identityTopUpTransition
+   * @return {ValidationResult}
+   */
+  async function validateIdentityTopUpTransitionData(identityTopUpTransition) {
+    const result = new ValidationResult();
+
+    // Check if identity with such id already exists
+    const identityId = identityTopUpTransition.getIdentityId();
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js:27]: Please use `validateIdentityExistence`

File: lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/**
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+ * @typedef {Object} RawIdentityTopUpTransition
+ * @extends AbstractStateTransition
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js]: Fixed

File: lib/stateTransition/RawStateTransition.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+/**
+ * @typedef RawStateTransition
+ * @property {number} protocolVersion
```
Comment (shumkov) [inline lib/stateTransition/RawStateTransition.js:3]: Let’s move it to AbstractStateTransition ?

File: lib/stateTransition/StateTransitionFacade.js [hunk 10]
```diff
@@ -107,6 +114,8 @@ class StateTransitionFacade {
       validatePublicKeys,
     );
 
+    const validateIdentityTopUpSTStructure = validateIdentityTopUpSTStructureFactory();
```
Comment (shumkov) [inline lib/stateTransition/StateTransitionFacade.js:117]: It seems we don’t need the factory :)

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js (hunk 2 above)
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpSTDataFactory.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+ * @property {number} protocolVersion
+ * @property {number} type
+ * @property {string} lockedOutPoint
+ * @property {string} identityId
+ * @property {number|null} signaturePublicKeyId
```
[the remark this answers is not in the document]
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/RawIdentityTopUpTransitionInterface.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js (hunk 6 above)
Comment (antouhou) [reply lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js]: Fixed

File: lib/stateTransition/StateTransitionFacade.js (hunk 10 above)
Comment (antouhou) [reply lib/stateTransition/StateTransitionFacade.js:117]: Haha true

File: lib/stateTransition/RawStateTransition.js (hunk 9 above)
Comment (antouhou) [reply lib/stateTransition/RawStateTransition.js:3]: Do you mean the contents of the file? Sure

Comment (antouhou) [reply lib/stateTransition/RawStateTransition.js]: Fixed

File: lib/stateTransition/StateTransitionFacade.js (hunk 10 above)
Comment (antouhou) [reply lib/stateTransition/StateTransitionFacade.js]: Fixed

File: lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+  /**
+   * @typedef validateIdentityTopUpTransitionData
+   * @param {IdentityTopUpTransition} identityTopUpTransition
+   * @return {ValidationResult}
+   */
+  async function validateIdentityTopUpTransitionData(identityTopUpTransition) {
+    const result = new ValidationResult();
+
+    result.merge(await validateIdentityExistence(identityTopUpTransition.getIdentityId()));
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.js:27]: Let's add some newlines for readability

File: lib/stateTransition/AbstractStateTransition.js [hunk 13]
```diff
@@ -13,6 +13,14 @@ const { encode } = require('../util/serializer');
 
 const calculateStateTransitionFee = require('./calculateStateTransitionFee');
 
+/**
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransition.js:16]: Let's move it under the class.

File: lib/stateTransition/AbstractStateTransition.js [hunk 14]
```diff
@@ -13,6 +13,14 @@ const { encode } = require('../util/serializer');
...
+ * @typedef RawStateTransition
+ * @property {number} protocolVersion
+ * @property {number} type
+ * @property {number|null} signaturePublicKeyId
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransition.js:20]: We don't have this property here.

File: lib/stateTransition/StateTransitionFacade.js [hunk 15]
```diff
@@ -120,6 +127,10 @@ class StateTransitionFacade {
         validationFunction: validateIdentityCreateSTStructure,
         schema: identityCreateTransitionSchema,
       },
+      [stateTransitionTypes.IDENTITY_TOP_UP]: {
+        validationFunction: validateIdentityTopUpSTStructure,
```
Comment (shumkov) [inline lib/stateTransition/StateTransitionFacade.js:131]: validateIdentityTopUpTransitionStructure

File: lib/stateTransition/StateTransitionFacade.js [hunk 16]
```diff
@@ -148,6 +159,11 @@ class StateTransitionFacade {
       validateLockTransaction,
     );
 
+    const validateIdentityTopUpSTData = validateIdentityTopUpSTDataFactory(
```
Comment (shumkov) [inline lib/stateTransition/StateTransitionFacade.js:162]: validateIdentityTopUpTransitionDataFactory

File: lib/stateTransition/AbstractStateTransition.js (hunk 14 above)
Comment (antouhou) [reply lib/stateTransition/AbstractStateTransition.js]: Fixed

File: lib/stateTransition/AbstractStateTransition.js (hunk 13 above)
Comment (antouhou) [reply lib/stateTransition/AbstractStateTransition.js]: Fixed

File: lib/stateTransition/StateTransitionFacade.js (hunk 15 above)
Comment (antouhou) [reply lib/stateTransition/StateTransitionFacade.js:131]: Fixed

File: lib/stateTransition/StateTransitionFacade.js (hunk 16 above)
Comment (antouhou) [reply lib/stateTransition/StateTransitionFacade.js]: Fixed

File: lib/test/fixtures/getIdentityTopUpSTFixture.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
+const IdentityTopUpTransition = require('../../identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition');
+
+const stateTransitionTypes = require('../../stateTransition/stateTransitionTypes');
+
+/**
+ *
+ * @return {IdentityTopUpTransition}
+ */
+module.exports = function getIdentityTopUpSTFixture() {
```
Comment (shumkov) [inline lib/test/fixtures/getIdentityTopUpSTFixture.js:9]: getIdentityTopUpTransitionFixture

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js [hunk 18]
```diff
@@ -0,0 +1,82 @@
+const getIdentityTopUpSTFixture = require('../../../../../lib/test/fixtures/getIdentityTopUpSTFixture');
+
+const { expectValidationError } = require(
+  '../../../../../lib/test/expect/expectError',
+);
+
+const validateIdentityTopUpSTDataFactory = require(
```
Comment (shumkov) [inline test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js:7]: Transition

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    expect(error.getIdentityId()).to.be.equal(stateTransition.getIdentityId());
+  });
+
+  it('should return valid result if state transition is valid', async () => {
+    const result = await validateIdentityTopUpSTData(stateTransition);
+
+    expect(result.isValid()).to.be.true();
+  });
+
+
```
Comment (shumkov) [inline test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js:67]: Extra newline

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionStructure.spec.js [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+  it('should pass valid raw state transition', () => {
+    const result = validateIdentityTopUpTransitionStructure(rawStateTransition);
+
+    expect(result.isValid()).to.be.true();
+  });
+
+  it('should pass valid state transition', () => {
+    const result = validateIdentityTopUpTransitionStructure(
+      new IdentityTopUpTransition(rawStateTransition),
```
Comment (shumkov) [inline test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionStructure.spec.js:29]: We could pass just raw thing here

File: test/unit/identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition.spec.js [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+
+describe('IdentityTopUpTransition', () => {
+  let rawStateTransition;
+  let stateTransition;
+  let hashMock;
+  let signerMock;
+  let IdentityTopUpTransition;
+
+  beforeEach(function beforeEach() {
+    rawStateTransition = {
```
Comment (shumkov) [inline test/unit/identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition.spec.js:15]: Use fixture?

File: test/unit/identity/stateTransitions/identityTopUpTransition/applyIdentityTopUpTransitionFactory.spec.js [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+    stateTransition = getIdentityTopUpSTFixture();
+    applyIdentityTopUpTransition = applyIdentityTopUpTransitionFactory(
+      stateRepositoryMock,
+      getLockedTransactionOutputMock,
+    );
+  });
+
+  it('should store identity created from state transition', async () => {
+    const balanceBeforeTopUp = identity.balance;
```
Comment (shumkov) [inline test/unit/identity/stateTransitions/identityTopUpTransition/applyIdentityTopUpTransitionFactory.spec.js:40]: getBalance()?

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionStructure.spec.js (hunk 20 above)
Comment (antouhou) [reply test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionStructure.spec.js:29]: The test above does this. This one is exactly to test the case when instnace is passed

File: lib/test/fixtures/getIdentityTopUpSTFixture.js (hunk 17 above)
Comment (antouhou) [reply lib/test/fixtures/getIdentityTopUpSTFixture.js]: Fixed

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js (hunk 19 above)
Comment (antouhou) [reply test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js]: Fixed

File: test/unit/identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition.spec.js (hunk 21 above)
Comment (antouhou) [reply test/unit/identity/stateTransitions/identityTopUpTransition/IdentityTopUpTransition.spec.js]: Fixed

File: test/unit/identity/stateTransitions/identityTopUpTransition/applyIdentityTopUpTransitionFactory.spec.js (hunk 22 above)
Comment (antouhou) [reply test/unit/identity/stateTransitions/identityTopUpTransition/applyIdentityTopUpTransitionFactory.spec.js]: Fixed

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js (hunk 18 above)
Comment (antouhou) [reply test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js]: Fixed

File: test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+  let validateLockTransactionMock;
+  let validateIdentityExistenceMock;
+
+  beforeEach(function beforeEach() {
+    const privateKey = 'af432c476f65211f45f48f1d42c9c0b497e56696aa1736b40544ef1a496af837';
+    stateRepositoryMock = createStateRepositoryMock(this.sinonSandbox);
+    validateLockTransactionMock = this.sinonSandbox.stub().returns(new ValidationResult());
+    validateIdentityExistenceMock = this.sinonSandbox.stub().resolves(new ValidationResult());
+
+    validateIdentityTopUpSTData = validateIdentityTopUpTransitionDataFactory(
```
Comment (shumkov) [inline test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js:32]: `validateIdentityTopUpTransitionData`

Comment (antouhou) [reply test/integration/identity/stateTransition/identityTopUpTransition/validateIdentityTopUpTransitionDataFactory.spec.js]: Fixed

**Pull request merged**